### ISSUE TITLE
Calling uno within a script fails

### ISSUE THREAD

**Issue by keith-turner** [apache/fluo-uno#185]
There is something about calling uno in a script like the following ...

```bash
cat testClasses.txt | while read test_class; do
  uno setup accumulo
done

```  

... that causes setup to fail with errors like the following :

```
/home/ubuntu/uno/bin/impl/setup-hadoop.sh: line 27: echo: write error: Bad file descriptor
/home/ubuntu/uno/bin/impl/setup-zookeeper.sh: line 26: echo: write error: Bad file descriptor
/home/ubuntu/uno/bin/impl/setup-accumulo.sh: line 33: echo: write error: Bad file descripto
```

I had to remove the file redirection for the setup scripts to see these errors.

**Comment by keith-turner** [comment]
Line 27 of setup-hadoop.sh is the following :

```
echo >&0 "Setting up Apache Hadoop at $HADOOP_PREFIX"
```

**Comment by keith-turner** [comment]
Calling `uno setup` in the script outside the while loop works fine.   I suspect it has something to do with the redirection for the loop.  Maybe this does something to stdin that borks `echo >&0`.  Not sure.

**Comment by ctubbsii** [comment]
I'm inclined to call this a bug in the caller... not in the uno script. It can be worked around by avoiding the pipe in the `| while read`, which hijacks file descriptors and keeps them in use during the loop.

For example, this fails:

```bash
{ for x in "a 1" "b 2" "c 3"; do echo $x; done; } | while read y; do echo >&0 "->$y"; done
```

But reading the input into an array first, and then looping over the array works fine:

```bash
readarray -t z < <(for x in "a 1" "b 2" "c 3"; do echo $x; done;); for y in "${z[@]}"; do echo >&0 "->$y"; done
```

In the above examples, the I'm using the `for x` loop to simulate an input file.

**Comment by keith-turner** [comment]
> I'm inclined to call this a bug in the caller.

I am not sure if that while pattern is buggy because I am not too familiar with bash.  However it was the first thing I found with a google search on stackoverflow.   Writing to stdin seems a bit buggy to me.  Debugging this was very painful because for some reason when this failed the logs were empty.  It was not until I removed the redirection from the scripts that I saw the error messages.  It took me a while to narrow the failure down to the point of knowing I needed to remove that redirection.

Would it be possible to attempt the write to stdin and if it fails just not print anything?

I experimented with the following changes.  For some reason the message about setting up Accumulo showed up but not the ones about Hadoop and ZK.  Even if this worked, not sure that taking over a random file descriptor is good.

```diff
diff --git a/bin/impl/setup-accumulo.sh b/bin/impl/setup-accumulo.sh
index ec179ad..f239df1 100755
--- a/bin/impl/setup-accumulo.sh
+++ b/bin/impl/setup-accumulo.sh
@@ -30,7 +30,7 @@ pkill -f accumulo.start
 # stop if any command fails
 set -e
 
-echo >&0 "Setting up Apache Accumulo at $ACCUMULO_HOME"
+echo >&5 "Setting up Apache Accumulo at $ACCUMULO_HOME"
 
 rm -rf "$INSTALL"/accumulo-*
 rm -f "$ACCUMULO_LOG_DIR"/*
diff --git a/bin/impl/setup-hadoop.sh b/bin/impl/setup-hadoop.sh
index d8a81ff..720ff77 100755
--- a/bin/impl/setup-hadoop.sh
+++ b/bin/impl/setup-hadoop.sh
@@ -24,7 +24,7 @@ pkill -f hadoop.yarn
 # stop if any command fails
 set -e
 
-echo >&0 "Setting up Apache Hadoop at $HADOOP_PREFIX"
+echo >&5 "Setting up Apache Hadoop at $HADOOP_PREFIX"
 
 rm -rf "$INSTALL"/hadoop-*
 rm -f "$HADOOP_LOG_DIR"/*
diff --git a/bin/impl/setup-zookeeper.sh b/bin/impl/setup-zookeeper.sh
index 9a6ef37..0814b83 100755
--- a/bin/impl/setup-zookeeper.sh
+++ b/bin/impl/setup-zookeeper.sh
@@ -23,7 +23,7 @@ pkill -f QuorumPeerMain
 # stop if any command fails
 set -e
 
-echo >&0 "Setting up Apache ZooKeeper at $ZOOKEEPER_HOME"
+echo >&5 "Setting up Apache ZooKeeper at $ZOOKEEPER_HOME"
 rm -rf "$INSTALL"/zookeeper-*
 rm -f "$ZOO_LOG_DIR"/*
 mkdir -p "$ZOO_LOG_DIR"
diff --git a/bin/impl/util.sh b/bin/impl/util.sh
index ac1a4fa..df1735b 100755
--- a/bin/impl/util.sh
+++ b/bin/impl/util.sh
@@ -57,5 +57,5 @@ function run_setup_script() {
   local L_DIR; L_DIR="$LOGS_DIR/setup"
   mkdir -p "$L_DIR"
   shift
-  "$UNO_HOME/bin/impl/setup-$SCRIP.sh" "$@" 1>"$L_DIR/$SCRIP.stdout" 2>"$L_DIR/$SCRIP.stderr"
+  "$UNO_HOME/bin/impl/setup-$SCRIP.sh" "$@" 5>&1 1>"$L_DIR/$SCRIP.stdout" 2>"$L_DIR/$SCRIP.stderr"
 }
```

**Comment by keith-turner** [comment]
I got things to work with redirecting stdout to another file descriptor.   The place where I previously did the redirecting was called multiple times.

```diff
diff --git a/bin/impl/setup-accumulo.sh b/bin/impl/setup-accumulo.sh
index ec179ad..f239df1 100755
--- a/bin/impl/setup-accumulo.sh
+++ b/bin/impl/setup-accumulo.sh
@@ -30,7 +30,7 @@ pkill -f accumulo.start
 # stop if any command fails
 set -e
 
-echo >&0 "Setting up Apache Accumulo at $ACCUMULO_HOME"
+echo >&5 "Setting up Apache Accumulo at $ACCUMULO_HOME"
 
 rm -rf "$INSTALL"/accumulo-*
 rm -f "$ACCUMULO_LOG_DIR"/*
diff --git a/bin/impl/setup-hadoop.sh b/bin/impl/setup-hadoop.sh
index d8a81ff..720ff77 100755
--- a/bin/impl/setup-hadoop.sh
+++ b/bin/impl/setup-hadoop.sh
@@ -24,7 +24,7 @@ pkill -f hadoop.yarn
 # stop if any command fails
 set -e
 
-echo >&0 "Setting up Apache Hadoop at $HADOOP_PREFIX"
+echo >&5 "Setting up Apache Hadoop at $HADOOP_PREFIX"
 
 rm -rf "$INSTALL"/hadoop-*
 rm -f "$HADOOP_LOG_DIR"/*
diff --git a/bin/impl/setup-zookeeper.sh b/bin/impl/setup-zookeeper.sh
index 9a6ef37..0814b83 100755
--- a/bin/impl/setup-zookeeper.sh
+++ b/bin/impl/setup-zookeeper.sh
@@ -23,7 +23,7 @@ pkill -f QuorumPeerMain
 # stop if any command fails
 set -e
 
-echo >&0 "Setting up Apache ZooKeeper at $ZOOKEEPER_HOME"
+echo >&5 "Setting up Apache ZooKeeper at $ZOOKEEPER_HOME"
 rm -rf "$INSTALL"/zookeeper-*
 rm -f "$ZOO_LOG_DIR"/*
 mkdir -p "$ZOO_LOG_DIR"
diff --git a/bin/uno b/bin/uno
index 94c74ab..a6c4394 100755
--- a/bin/uno
+++ b/bin/uno
@@ -40,6 +40,7 @@ fetch)
   fi
        ;;
 setup)
+  exec 5>&1
   [[ -n $LOGS_DIR ]] && rm -f "$LOGS_DIR"/setup/*.std{out,err}
   echo "Beginning setup (detailed logs in $LOGS_DIR/setup)..."
   case "$2" in
```

With the above change I see the following. Still not sure if redirecting fd 5 is a good thing to do.

```
$ ./bin/uno setup accumulo
Beginning setup (detailed logs in /home/ubuntu/git/uno/install/logs/setup)...
Setting up Apache Hadoop at /home/ubuntu/git/uno/install/hadoop-2.8.4
Setting up Apache ZooKeeper at /home/ubuntu/git/uno/install/zookeeper-3.4.12
Setting up Apache Accumulo at /home/ubuntu/git/uno/install/accumulo-2.0.0-SNAPSHOT
Setup complete.
```

**Comment by ctubbsii** [comment]
Unless rewritten, the built-in initial file descriptors, `0`, `1`, and `2`, all point to the same console device by default. You may be able to create additional ones which also point to the same device. Using `0` is not necessarily any different (or more buggy) than using any other file descriptor aliased to that same device. The tricky part is finding one that isn't going to be messed with (rewritten) by other things during routine use.

Here are reasons why `fd/0` was suitable:
1. Like `1` and `2`, it exists by default,
1. It may not always be possible to create new ones (might depend on platform or permissions or system limits),
1. Creating new file descriptors may require adding complexity to ensure the new ones are cleaned up when no longer needed, even in the case of failure,
1. `0` is rarely, if ever, rewritten by `>`, and thus expected to always point to the console, and
1. The pipe `|` does rewrite `fd/0` routinely, but since the uno script doesn't take any input, there wasn't any reason to expect that anybody would rewrite `fd/0` with a pipe.

Using a different arbitrary one, like `5` is perfectly fine, as long as:

1. You don't expect anybody else to rewrite it,
1. You clean it up when done with it,
1. You assume the user can arbitrarily create new file descriptors wherever the script is executed, and
1. You accept the limitation that whatever number you choose isn't in use/rewritten by the caller (like `fd/0` was when the caller used a pipe).

It may be simpler to just avoid rewriting `fd/0` by taking the uno script outside of a loop with a pipe involved, though, and accept that limitation. Either way works for me, but hopefully you understand the issue in greater detail now. :smiley_cat:

**Comment by ctubbsii** [comment]
So, I just realized `readarray` is a bash 4 thing, so it likely won't work on Macs, anyway; it's still possible to read stuff into an array first, and then loop the array when calling uno, but... it's a few extra lines, and using a new file descriptor (if it works on a Mac) is probably better.

**Comment by keith-turner** [comment]
I am thinking of introducing a `printToConsole` function in the code that replaces the existing `echo >&0` commands.
